app/src/main/java/com/hedvig/app/feature/addressautocompletion/ui/AddressAutoCompleteViewModel.kt

```diff
@@ -0,0 +1,97 @@
+package com.hedvig.app.feature.addressautocompletion.ui
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.hedvig.app.feature.addressautocompletion.data.AddressAutoCompleteResults
+import com.hedvig.app.feature.addressautocompletion.data.GetDanishAddressAutoCompletionUseCase
+import com.hedvig.app.feature.addressautocompletion.model.DanishAddress
+import com.hedvig.app.feature.addressautocompletion.model.DanishAddressInput
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+
+@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
+class AddressAutoCompleteViewModel(
+    initialAddress: DanishAddress?,
+    getDanishAddressAutoCompletionUseCase: GetDanishAddressAutoCompletionUseCase,
+) : ViewModel() {
+
+    private val currentInput: MutableStateFlow<DanishAddressInput> = MutableStateFlow(
+        DanishAddressInput.fromDanishAddress(initialAddress)
+    )
+    private val queryResults: Flow<List<DanishAddress>> = currentInput
+        .debounce(50)
+        .mapLatest { input ->
+            getDanishAddressAutoCompletionUseCase.invoke(input)
+                .fold(
+                    { emptyList() },
+                    AddressAutoCompleteResults::resultList
+                )
+        }
+        .onStart { emit(emptyList()) }
+
+    private val checkSelectionChannel: Channel<Pair<DanishAddress, List<DanishAddress>>> = Channel(Channel.UNLIMITED)
+    val resultEvent = checkSelectionChannel.receiveAsFlow()
+        .mapLatest { (selectedAddress, addressList) ->
+            val address = getFinalResultOrNull(selectedAddress, addressList) ?: return@mapLatest null
+            AddressAutoCompleteEvent.Selection(address)
+        }
+        .filterNotNull()
+
+    val viewState: StateFlow<AddressAutoCompleteViewState> =
+        combine(currentInput, queryResults) { input, results ->
```

Comment:
Are we handling errors here from queryResults?

Reply:
What we're doing is that we're simply showing no results (since we get none) and the "I can't find my address" red option shows.
Thing is, there is no design on how to handle when there is no network access in this screen, and this would be the only case where there would be an error from "queryResults". I think this is reasonable if the other clients are doing the same. In fact the web is doing the same, but not showing the "Can't find my address" option at all.

Reply:
I see, that's fine. We should generally be better at handling errors though - no internet connection is not the only thing that can go wrong here. The API can change, headers can break etc. Showing the member what has gone wrong is almost always better UX than not showing anything at all IMO!

Reply:
Yeah, do you want to bring this up sometime? Whenever there's places like this where error-handling isn't obvious and I can't come up with something on the fly myself, I've learned to just not do it because otherwise I spend way too much time and it ends up looking hella bad (I am horrible at designing stuff 😂).

So if going forward, if we want to approach it in a way where we bring it up early and make something about it with our designers before we actually implement it then we'd avoid situations like this. It's also been released like this in iOS for ~8 months now so I guess it's been accepted to be that way.